From 3cee9d3c5f43c4aeb48b1483a54f8694361050e9 Mon Sep 17 00:00:00 2001
From: Hyun Kwon <hyun.kwon@xilinx.com>
Date: Wed, 11 Mar 2020 14:57:56 -0700
Subject: [PATCH 1280/1852] v4l: xilinx: vipp: move xvip_entity_start_stop() to
 xilinx-vipp

commit 239b4d6cbe37660c9e1aa475615e81b77ccfd948 from
https://github.com/Xilinx/linux-xlnx.git

This is a preparation to do the graph level stream on/off handling
in xilinx-vipp. Rename an argument name to align: 'start' to 'on'.

Signed-off-by: Hyun Kwon <hyun.kwon@xilinx.com>
Reviewed-by: Vishal Sagar <vishal.sagar@xilinx.com>
Reviewed-by: Dylan Yip <dylan.yip@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/media/platform/xilinx/xilinx-dma.c  | 54 -------------------
 drivers/media/platform/xilinx/xilinx-vipp.c | 59 ++++++++++++++++++++-
 drivers/media/platform/xilinx/xilinx-vipp.h |  4 +-
 3 files changed, 59 insertions(+), 58 deletions(-)

diff --git a/drivers/media/platform/xilinx/xilinx-dma.c b/drivers/media/platform/xilinx/xilinx-dma.c
index 441e3d56f499..3fb95ef3e8b7 100644
--- a/drivers/media/platform/xilinx/xilinx-dma.c
+++ b/drivers/media/platform/xilinx/xilinx-dma.c
@@ -95,60 +95,6 @@ static int xvip_dma_verify_format(struct xvip_dma *dma)
  * Pipeline Stream Management
  */
 
-static int xvip_entity_start_stop(struct xvip_composite_device *xdev,
-				  struct media_entity *entity, bool start)
-{
-	struct v4l2_subdev *subdev;
-	bool is_streaming;
-	int ret = 0;
-
-	dev_dbg(xdev->dev, "%s entity %s\n",
-		start ? "Starting" : "Stopping", entity->name);
-	subdev = media_entity_to_v4l2_subdev(entity);
-
-	/* This is to maintain list of stream on/off devices */
-	is_streaming = xvip_subdev_set_streaming(xdev, subdev, start);
-
-	/*
-	 * start or stop the subdev only once in case if they are
-	 * shared between sub-graphs
-	 */
-	if (start && !is_streaming) {
-		/* power-on subdevice */
-		ret = v4l2_subdev_call(subdev, core, s_power, 1);
-		if (ret < 0 && ret != -ENOIOCTLCMD) {
-			dev_err(xdev->dev,
-				"s_power on failed on subdev\n");
-			xvip_subdev_set_streaming(xdev, subdev, 0);
-			return ret;
-		}
-
-		/* stream-on subdevice */
-		ret = v4l2_subdev_call(subdev, video, s_stream, 1);
-		if (ret < 0 && ret != -ENOIOCTLCMD) {
-			dev_err(xdev->dev,
-				"s_stream on failed on subdev\n");
-			v4l2_subdev_call(subdev, core, s_power, 0);
-			xvip_subdev_set_streaming(xdev, subdev, 0);
-		}
-	} else if (!start && is_streaming) {
-		/* stream-off subdevice */
-		ret = v4l2_subdev_call(subdev, video, s_stream, 0);
-		if (ret < 0 && ret != -ENOIOCTLCMD) {
-			dev_err(xdev->dev,
-				"s_stream off failed on subdev\n");
-			xvip_subdev_set_streaming(xdev, subdev, 1);
-		}
-
-		/* power-off subdevice */
-		ret = v4l2_subdev_call(subdev, core, s_power, 0);
-		if (ret < 0 && ret != -ENOIOCTLCMD)
-			dev_err(xdev->dev,
-				"s_power off failed on subdev\n");
-	}
-
-	return ret;
-}
 
 /**
  * xvip_pipeline_start_stop - Start ot stop streaming on a pipeline
diff --git a/drivers/media/platform/xilinx/xilinx-vipp.c b/drivers/media/platform/xilinx/xilinx-vipp.c
index 99926d2d363c..96c4ea273e76 100644
--- a/drivers/media/platform/xilinx/xilinx-vipp.c
+++ b/drivers/media/platform/xilinx/xilinx-vipp.c
@@ -204,8 +204,8 @@ xvip_graph_find_dma(struct xvip_composite_device *xdev, unsigned int port)
  * Return: streaming status (true or false) if successful or warn_on if subdev
  * is not present and return false
  */
-bool xvip_subdev_set_streaming(struct xvip_composite_device *xdev,
-			       struct v4l2_subdev *subdev, bool enable)
+static bool xvip_subdev_set_streaming(struct xvip_composite_device *xdev,
+				      struct v4l2_subdev *subdev, bool enable)
 {
 	struct xvip_graph_entity *entity;
 	struct v4l2_async_subdev *asd;
@@ -224,6 +224,61 @@ bool xvip_subdev_set_streaming(struct xvip_composite_device *xdev,
 	return false;
 }
 
+int xvip_entity_start_stop(struct xvip_composite_device *xdev,
+			   struct media_entity *entity, bool on)
+{
+	struct v4l2_subdev *subdev;
+	bool is_streaming;
+	int ret = 0;
+
+	dev_dbg(xdev->dev, "%s entity %s\n",
+		on ? "Starting" : "Stopping", entity->name);
+	subdev = media_entity_to_v4l2_subdev(entity);
+
+	/* This is to maintain list of stream on/off devices */
+	is_streaming = xvip_subdev_set_streaming(xdev, subdev, on);
+
+	/*
+	 * start or stop the subdev only once in case if they are
+	 * shared between sub-graphs
+	 */
+	if (on && !is_streaming) {
+		/* power-on subdevice */
+		ret = v4l2_subdev_call(subdev, core, s_power, 1);
+		if (ret < 0 && ret != -ENOIOCTLCMD) {
+			dev_err(xdev->dev,
+				"s_power on failed on subdev\n");
+			xvip_subdev_set_streaming(xdev, subdev, 0);
+			return ret;
+		}
+
+		/* stream-on subdevice */
+		ret = v4l2_subdev_call(subdev, video, s_stream, 1);
+		if (ret < 0 && ret != -ENOIOCTLCMD) {
+			dev_err(xdev->dev,
+				"s_stream on failed on subdev\n");
+			v4l2_subdev_call(subdev, core, s_power, 0);
+			xvip_subdev_set_streaming(xdev, subdev, 0);
+		}
+	} else if (!on && is_streaming) {
+		/* stream-off subdevice */
+		ret = v4l2_subdev_call(subdev, video, s_stream, 0);
+		if (ret < 0 && ret != -ENOIOCTLCMD) {
+			dev_err(xdev->dev,
+				"s_stream off failed on subdev\n");
+			xvip_subdev_set_streaming(xdev, subdev, 1);
+		}
+
+		/* power-off subdevice */
+		ret = v4l2_subdev_call(subdev, core, s_power, 0);
+		if (ret < 0 && ret != -ENOIOCTLCMD)
+			dev_err(xdev->dev,
+				"s_power off failed on subdev\n");
+	}
+
+	return ret;
+}
+
 static int xvip_graph_build_dma(struct xvip_composite_device *xdev)
 {
 	u32 link_flags = MEDIA_LNK_FL_ENABLED;
diff --git a/drivers/media/platform/xilinx/xilinx-vipp.h b/drivers/media/platform/xilinx/xilinx-vipp.h
index 24934d57529b..c8ce81853945 100644
--- a/drivers/media/platform/xilinx/xilinx-vipp.h
+++ b/drivers/media/platform/xilinx/xilinx-vipp.h
@@ -41,7 +41,7 @@ struct xvip_composite_device {
 	struct mutex lock; /* lock to protect xvip pipeline instance */
 };
 
-bool xvip_subdev_set_streaming(struct xvip_composite_device *xdev,
-			       struct v4l2_subdev *subdev, bool enable);
+int xvip_entity_start_stop(struct xvip_composite_device *xdev,
+			   struct media_entity *entity, bool on);
 
 #endif /* __XILINX_VIPP_H__ */
-- 
2.31.1

